### ISSUE TITLE
Rate limiting middleware - Statistics about rate limiters

### ISSUE THREAD

**Issue by maartenba** [dotnet/aspnetcore#44140]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

The ASP.NET Core rate limiting middleware is great, but "limited" in terms of what you an communicate with your users. Let's start with some code that you can write today in .NET 7:

```csharp
builder.Services.AddRateLimiter(options =>
{
    options.OnRejected = async (context, token) =>
    {
        context.HttpContext.Response.StatusCode = 429;
        if (context.Lease.TryGetMetadata(MetadataName.RetryAfter, out var retryAfter))
        {
            await context.HttpContext.Response.WriteAsync(
                $"Too many requests. Please try again after {retryAfter.TotalMinutes} minute(s). " +
                $"Read more about our rate limits at https://example.org/docs/ratelimiting.", cancellationToken: token);
        }
        else
        {
            await context.HttpContext.Response.WriteAsync(
                "Too many requests. Please try again later. " +
                "Read more about our rate limits at https://example.org/docs/ratelimiting.", cancellationToken: token);
        }
    };

    // ...
});
```

When rate limits are triggered, a response is returned that tells the user they are rate limited, where to find more information, and (if the `MetadataName.RetryAfter` data is available), when to retry.

This is quite limited. There's no access to which rate limiter fired, and what its statistics are.

Additionally, the current `RateLimitLease` is only accessible when rate limiting is fired - not for every request. If you would want to return statistics about your limits (e.g. [like GitHub does](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limit-http-headers)), you'll find it impossible to get statistics about the current lease in a custom middleware that can write out these additional headers.

### Describe the solution you'd like

Here's a middleware that has access to some of data that I'd want to access:

```csharp
public class RateLimitStatisticsMiddleware
{
    private readonly RequestDelegate _next;
    private readonly IOptions<RateLimiterOptions> _options;

    public RateLimitStatisticsMiddleware(RequestDelegate next, IOptions<RateLimiterOptions> options)
    {
        _next = next;
        _options = options;
    }

    public Task Invoke(HttpContext context)
    {
        // Note: This should also work on endpoint limiters, but those are not available.
        // There is no current "rate limit context" of sorts. Including the policy name etc.
        var globalLimiter = _options.Value.GlobalLimiter;
        if (globalLimiter != null)
        {
            var statistics = globalLimiter.GetStatistics(context);
            if (statistics != null)
            {
                // Note: It would be great to be able to get the TokenLimit from the "current rate limiter context"
                context.Response.Headers.Add("X-Rate-Limit-Limit", "20");

                // Note: It would be great to be able to get the Window from the "current rate limiter context"
                context.Response.Headers.Add("X-Rate-Limit-Reset", DateTimeOffset.UtcNow.ToString("O"));

                // This one is there today
                context.Response.Headers.Add("X-Rate-Limit-Remaining", statistics.CurrentAvailablePermits.ToString());
            }
        }

        return _next(context);
    }
}
```

The dream scenario for a better ASP.NET Core rate limiter middleware would be to:

* Have access to more statistics about the rate limiter (which policy fired, what partition, what are the limiter's options so I can emit a message that says "you can only make 10 requests per minute", ...) in the rejected callbacks.
* Have a feature on the current `HttpContext` that gives access to the current rate limit context, so these details can also be returned on successful requests, or added to telemetry.

### Additional context

Most rate limiters in `System.Threading.RateLimiting` don't provide additional statistics. This feature will need changes in both ASP.NET Core and the .NET framework.

**Comment by adityamandaleeka** [comment]
@BrennanConroy @wtgodbe @halter73

**Comment by wtgodbe** [comment]
I definitely think we should do this for dotnet 8

**Comment by cristipufu** [comment]
> Have a feature on the current HttpContext that gives access to the current rate limit context, so these details can also be returned on successful requests, or added to telemetry.

How would we handle this for multiple policies (eg global + endpoint)?

**Comment by maartenba** [comment]
How does the runtime handle such case? I assume the most strict is applied?

**Comment by japui-coder** [comment]
Rate limiting middleware - Statistics about rate limiters #44140

**Comment by MadL1me** [comment]
@maartenba you said: 

> Most rate limiters in System.Threading.RateLimiting don't provide additional statistics. This feature will need changes in both ASP.NET Core and the .NET framework.

Can you please explain on the process how to do this? Do I need to send link to this issue in PR to other repo? Where I can find BCL (with rate .net base rate limiting) repo on a github? Do I need somehow to contact devs via issue submitting, or should I right away submit a PR? It would be great to have answers to those

**Comment by maartenba** [comment]
@MadL1me The BCL classes are in https://github.com/dotnet/runtime/tree/main/src/libraries/System.Threading.RateLimiting

I am not Microsoft, so no idea how to do a PR across multiple repos/how to suggest such changes. @wtgodbe may be able to comment on that?

**Comment by wtgodbe** [comment]
We ingest changes from dotnet/runtime via automated dependency updates. If your change to the BCL base limiter classes requires API change, you'll need to open a [formal API proposal](https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.yml&title=%5BAPI+Proposal%5D%3A+). Otherwise you can just open PR and ask for review from myself, @BrennanConroy, and @Tratcher. Once your PR is merged into dotnet/runtime, after a few hours an automated dependency update PR should get opened in this repo (will look like https://github.com/dotnet/aspnetcore/pull/45448). Once that's merged, your changes from dotnet/runtime will be available in this repo, and you can open your follow-up PR here.

**Comment by anandsowm** [comment]
Yes, would be very useful feature. Looking at the different classes involved, there is no way to know how many permits have been issued so far for a given partition key. That would be good to have. Also is there a way to retain the count of leases issued between system restarts? It is right now stored in-memory, try to provide a way to plug in a persistence component.
